Firestore/core/src/firebase/firestore/util/executor_std.h

```diff
@@ -0,0 +1,283 @@
+ public:
+  // Entries are scheduled using absolute time.
+  using TimePoint =
+      std::chrono::time_point<std::chrono::system_clock, Duration>;
+
+  // Schedules an entry for the specified time due. `due` may be in the past.
+  void Push(const T& value, const TimePoint due) {
+    InsertPreservingOrder(Entry{value, due});
+  }
+  void Push(T&& value, const TimePoint due) {
+    InsertPreservingOrder(Entry{std::move(value), due});
+  }
+
+  // If the queue contains at least one entry for which the scheduled time is
+  // due now (according to the system clock), removes the entry which is the
+  // most overdue from the queue and returns it. If no entry is due, returns an
+  // empty `optional`.
+  absl::optional<T> PopIfDue() {
+    std::lock_guard<std::mutex> lock{mutex_};
+
+    if (HasDue()) {
+      return Extract(scheduled_.begin());
+    }
+    return {};
+  }
+
+  // Blocks until at least one entry is available for which the scheduled time
+  // is due now (according to the system clock), removes the entry which is the
+  // most overdue from the queue and returns it. The function will
+  // attempt to minimize both the waiting time and busy waiting.
+  T PopBlocking() {
+    std::unique_lock<std::mutex> lock{mutex_};
+
+    while (true) {
+      cv_.wait(lock, [this] { return !scheduled_.empty(); });
+
+      // To minimize busy waiting, sleep until either the nearest entry in the
+      // future either changes, or else becomes due.
+      const auto until = scheduled_.front().due;
+      cv_.wait_until(lock, until,
+                     [this, until] { return scheduled_.front().due != until; });
+      // There are 3 possibilities why `wait_until` has returned:
+      // - `wait_until` has timed out, in which case the current time is at
+      //   least `until`, so there must be an overdue entry;
+      // - a new entry has been added which comes before `until`. It must be
+      //   either overdue (in which case `HasDue` will break the cycle), or else
+      //   `until` must be reevaluated (on the next iteration of the loop);
+      // - `until` entry has been removed. This means `until` has to be
+      //   reevaluated, similar to #2.
+
+      if (HasDue()) {
+        return Extract(scheduled_.begin());
+      }
+    }
+  }
+
+  bool empty() const {
+    std::lock_guard<std::mutex> lock{mutex_};
+    return scheduled_.empty();
+  }
+
+  size_t size() const {
+    std::lock_guard<std::mutex> lock{mutex_};
+    return scheduled_.size();
+  }
+
+  // Removes the first entry satisfying predicate from the queue and returns it.
+  // If no such entry exists, returns an empty `optional`. Predicate is applied
+  // to entries in order according to their scheduled time.
+  //
+  // Note that this function doesn't take into account whether the removed entry
+  // is past its due time.
+  template <typename Pred>
+  absl::optional<T> RemoveIf(const Pred pred) {
+    std::lock_guard<std::mutex> lock{mutex_};
+
+    const auto found =
```

Comment:
Call me a luddite but a for loop would be clearer:

```c++
for (auto iter = scheduled_.begin(), end = scheduled_.end(); iter != end; ++iter) {
  if (pred(iter->value)) {
    return Extract(iter);
  }
}
return {};
```

It's shorter and more directly expresses the coupling between `pred` testing true and calling `Extract` on the iterator backing it.

Alternatively, because you have essentially the identical `std::find_if` below, maybe break that into a helper `Find(const Pred pred)` (that assumes the lock is held)?